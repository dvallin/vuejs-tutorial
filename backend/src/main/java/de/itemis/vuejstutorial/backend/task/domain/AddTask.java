package de.itemis.vuejstutorial.backend.task.domain;

import com.fasterxml.jackson.annotation.JsonProperty;

import javax.validation.constraints.NotBlank;

public class AddTask {
    @NotBlank(message = "Task title must not be blank!")
    public final String title;

    public AddTask(@JsonProperty("title") final String title) {
        this.title = title;
    }
}
