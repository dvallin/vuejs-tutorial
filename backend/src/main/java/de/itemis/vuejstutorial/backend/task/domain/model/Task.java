package de.itemis.vuejstutorial.backend.task.domain.model;

import org.springframework.data.annotation.Id;

import javax.validation.constraints.NotBlank;
import java.util.UUID;

public class Task {
    @Id
    public final UUID id;
    
    public final String title;

    public Task(final String title) {
        this.id = UUID.randomUUID();
        this.title = title;
    }
}
