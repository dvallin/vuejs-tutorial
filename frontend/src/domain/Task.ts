export enum TaskState {
  Open,
  Done
}

export class Task {
  title: string;
  state: TaskState;

  constructor(title: string) {
    this.title = title;
    this.state = TaskState.Open;
  }
}
