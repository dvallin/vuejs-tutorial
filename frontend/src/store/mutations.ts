
import {Task, TaskState} from "../domain/Task";

export enum Mutations {
  TaskAdded = 'taskAdded',
  TaskStateChanged = 'taskStateChanged',
  TaskNotFound = 'taskNotFound'
}

export interface TaskAdded {
  type: Mutations.TaskAdded;
  task: Task;
}
export interface TaskStateChanged {
  type: Mutations.TaskStateChanged;
  task: Task;
  newTaskState: TaskState;
}
export interface TaskNotFound {
  type: Mutations.TaskNotFound;
  taskTitle: string;
}

export const mutations = {
  [Mutations.TaskAdded] (state, payload: TaskAdded) {
    state.tasks.push(payload.task);
  },
  [Mutations.TaskStateChanged] (state, payload: TaskStateChanged) {
    payload.task.state = payload.newTaskState;
  },
  [Mutations.TaskNotFound] (state, payload: TaskStateChanged) {
    state.errors.push(payload);
  }
};
