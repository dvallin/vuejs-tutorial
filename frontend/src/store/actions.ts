import _ from 'lodash';

import * as TasksApi from "../api/tasks";

import {TaskState} from "../domain/Task";
import {Mutations} from "./mutations";

export enum Actions {
  AddTask = 'addTask',
  FetchTasks = 'fetchTasks',
  SetTaskComplete = 'setTaskComplete',
}

export interface AddTask {
  type: Actions.AddTask;
  title: string;
}

export interface FetchTasks {
  type: Actions.FetchTasks;
}

export interface SetTaskComplete {
  type: Actions.SetTaskComplete;
  title: string;
  complete: boolean;
}

export const actions = {
  async [Actions.AddTask] ({getters, commit}, {title}: AddTask): Promise<void> {
    try {
      const foundTask = getters.taskByTitle(title);
      if(_.isNil(foundTask)) {
        const task = await TasksApi.addTask(title);
        commit({ type: Mutations.AddTaskSucceeded, task });
      }
    } catch(error) {
      commit({type: Mutations.AddTaskFailed, title });
    }
  },
  async [Actions.FetchTasks] ({commit}, payload: FetchTasks): Promise<void> {
    try {
      const tasks = await TasksApi.fetchTasks();
      commit({type: Mutations.FetchTasksSucceeded, tasks});
    } catch(error) {
      commit({ type: Mutations.FetchTasksFailed, error: error.message });
    }
  },
  [Actions.SetTaskComplete] ({getters, commit}, payload: SetTaskComplete) {
    const task = getters.taskByTitle(payload.title);
    if(_.isNil(task)) {
      commit({
        type: Mutations.TaskNotFound,
        title: payload.title
      });
    } else {
      const newTaskState = payload.complete ? TaskState.Done : TaskState.Open;
      commit({
        type: Mutations.TaskStateChanged,
        task,
        newTaskState
      })
    }
  }
};
