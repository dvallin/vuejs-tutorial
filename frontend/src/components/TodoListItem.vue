<template>
  <div class="todo-list-item">
    <input type="checkbox" v-model="isComplete"/>
    <label>{{task.title}}</label>
  </div>
</template>

<script lang="ts">
  import {TaskState} from "../domain/Task";
  import {Actions} from "../store/actions";

  export default {
    name: 'TodoListItem',
    props: ['task'],

    data() {
      return { isComplete: this.task.state === TaskState.Done }
    },
    watch: {
      isComplete: 'handleCompleteClick'
    },
    methods: {
      handleCompleteClick() {
        this.$store.dispatch({
          type: Actions.SetTaskComplete,
          complete: this.isComplete,
          title: this.task.title
        });
      }
    }
  }
</script>
