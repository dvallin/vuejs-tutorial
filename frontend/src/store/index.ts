import * as Vuex from 'vuex';
import Vue from 'vue';

import {mutations} from './mutations';
import {actions} from "./actions";

import {Task} from "@/domain/Task";

export const initialState = () => {
  return {
    tasks: [],
    errors: []
  }
};

export const getters = {
  taskByTitle: (state, getters) => (title) => {
    return state.tasks.find(task => task.title === title)
  }
};

Vue.use(Vuex);
export default new Vuex.Store({
  state: initialState(),
  getters,
  mutations,
  actions
});
