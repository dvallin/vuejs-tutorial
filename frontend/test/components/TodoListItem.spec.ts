import TodoListItem from '../../src/components/TodoListItem'
import {Task, TaskState} from "../../src/domain/Task";
import store from '../../src/store'
import {mount} from 'vue-test-utils'
import {Actions} from "../../src/store/actions";

describe('TodoListItem.vue', () => {
  it('renders', () => {
    const wrapper = mount(TodoListItem, {
      propsData: {
        task: new Task("Some Title")
      }
    });
    expect(wrapper.element).toMatchSnapshot();
    expect(wrapper.find('label').text()).toBe("Some Title");
    expect(wrapper.find('input').hasAttribute("type", "checkbox")).toBeTruthy();
  });

  describe('checkbox', () => {
    it('is checked if task is done', () => {
      const task = new Task("Some Title");
      task.state = TaskState.Done;
      const wrapper = mount(TodoListItem, { propsData: { task } });
      expect(wrapper.find('input').element.checked).toBeTruthy();
    });

    it('is not checked if task is not done', () => {
      const task = new Task("Some Title");
      const wrapper = mount(TodoListItem, {  propsData: { task } });
      expect(wrapper.find('input').element.checked).toBeFalsy();
    });

    it('dispatches task state change', (done) => {
      const spy = jest.spyOn(store, 'dispatch');
      const givenTitle = "title";

      const wrapper = mount(TodoListItem, {
        store,
        propsData: {
          task: new Task(givenTitle)
        }
      });

      wrapper.find('input').trigger('click');

      wrapper.vm.$nextTick(() => {
        expect(spy).toHaveBeenCalledWith({
          type: Actions.SetTaskComplete,
          title: givenTitle,
          complete: true
        });
        done();
      });
    });
  });
});
