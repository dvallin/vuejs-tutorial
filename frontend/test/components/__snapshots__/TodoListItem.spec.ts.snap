// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TodoListItem.vue renders 1`] = `
<div
  class="todo-list-item"
>
  <input
    type="checkbox"
  />
   
  <label>
    Some Title
  </label>
</div>
`;
